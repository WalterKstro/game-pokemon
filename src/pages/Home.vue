<template>

    <div class="container-game" v-if="objectPokemonSelected">
        <Title/>
        <div class="container-main">
            <pokemon-image 
                :pokemonId="objectPokemonSelected.id" 
                :showImagePokemon="showImagePokemon"/>

            <pokemon-options 
                :arrayOptionsPokemons = "arrayOptions" 
                @select-pokemon="methodEmmitedOfPokemonOptions"/>
        </div>
        <pokemon-result  
            :message="messageResultGame"
            @new-game="tryOtherGame"/>
    </div>
    <bounce-loader 
        :loading="true" 
        v-else>
    </bounce-loader>

</template>

<script>
import PokemonImage from '@/components/PokemonImage.vue';
import PokemonOptions from '@/components/PokemonOptions.vue';
import Title from '@/components/Title.vue';
import PokemonResult from '@/components/PokemonResult.vue';
import getArrayRandomElement from '@/helpers/helpersArrays.js';
import BounceLoader from 'vue-spinner/src/BounceLoader.vue';

export default {
    name: 'Home',
    async created(){
        await this.getDataPokemon();
    },
    data() {
        return {
            arrayOptions : [],
            objectPokemonSelected : null,
            showImagePokemon : false,
            messageResultGame : null,
        }
    },
    components: {   
        PokemonImage,
        PokemonOptions,
        Title,
        PokemonResult,
        BounceLoader
    },
    methods : {
        async getDataPokemon(){
            this.arrayOptions = await getArrayRandomElement();
            this.objectPokemonSelected = this.arrayOptions[this.generateRandomNumber];
        },
        methodEmmitedOfPokemonOptions(id){
            this.showImagePokemon = true;
            if(this.objectPokemonSelected.id == id){
                this.messageResultGame = '¡Has ganado!';
            }else{
                this.messageResultGame = `¡Has perdido!, era ${this.objectPokemonSelected.name}`;
            }
        },
        async tryOtherGame(){
            this.arrayOptions = await getArrayRandomElement();
            this.objectPokemonSelected = this.arrayOptions[this.generateRandomNumber];
            this.showImagePokemon = false;
            this.messageResultGame = null;
        }
    },
    computed: {
        generateRandomNumber(){
            return Math.floor(Math.random() * this.arrayOptions.length);
        }
    }
}
</script>
<style>

@media screen and (min-width: 1024px){
    .container-main{
        display: flex;
        align-items: center;
        justify-content: space-evenly;
    }
}
.v-spinner{
    position:absolute;
    left:50%;
    transform: translateX(-50%);
}
</style>