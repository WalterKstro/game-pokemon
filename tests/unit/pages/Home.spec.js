import { mount, shallowMount } from "@vue/test-utils";
import Home from "@/pages/Home";

describe('Test suite for Home page', () => {
    let wrapper;
    let initialWrapper;
    const arrayPokemons = [
        { id: 196, name: 'espeon' },
        { id: 630, name: 'mandibuzz' },
        { id: 173, name: 'cleffa' },
        { id: 238, name: 'smoochum' }
    ];
    beforeEach(() => {
        wrapper = shallowMount(Home);
        initialWrapper = shallowMount(Home,{
            data(){
                return {
                    arrayOptions : arrayPokemons,
                    objectPokemonSelected : arrayPokemons[0],
                    showImagePokemon : false,
                    messageResultGame : null,
                }
            }
        });
    });

    test('Should call the getDataPokemon method, when the component is created', () => {
        // Un spy, es como un watch, que esta pendiente de la llamada a una funcion
        const spy = jest.spyOn(Home.methods, 'getDataPokemon');
        const wrapper = shallowMount(Home);

        expect(spy).toHaveBeenCalled();
    });

    test('Should create a correct snapshot of <Home/>',() => {
        const wrapper = mount(Home,{
            data(){
                return {
                    arrayOptions : arrayPokemons,
                    objectPokemonSelected : arrayPokemons[0],
                    showImagePokemon : true,
                    messageResultGame : null,
                }
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    })


    test('Should exist the component <PokemonOptions/>', () => {
        expect(initialWrapper.find('pokemon-options-stub').exists()).toBe(true);
    });
    test('Should exist the component <PokemonImage/>', () => {
        expect(initialWrapper.find('pokemon-image-stub').exists()).toBe(true);
    });
    test('Should exist the component <PokemonResult/>', () => {
        expect(initialWrapper.find('pokemon-result-stub').exists()).toBe(true);
    });

    test('Should the component <PokemonImage, has the attribute pokemonId with 196/>', () => {
        const objectProperties = initialWrapper.find('pokemon-image-stub').attributes();

        expect(objectProperties.hasOwnProperty('pokemonid')).toBeTruthy();
        expect(objectProperties.pokemonid).toBe('196');
    });

    test('Should the component <PokemonOptions, has the attribute  arrayoptionspokemons/>', () => {
        const objectProperties = initialWrapper.find('pokemon-options-stub').attributes();
        expect(objectProperties.hasOwnProperty('arrayoptionspokemons')).toBeTruthy();
    });

    test('Should execute the method methodEmmitedOfPokemonOptions, when the component <PokemonOptions>, emits the event "emitPokemonSelected"', () => {
        expect(initialWrapper.vm.messageResultGame).toBeNull();

        initialWrapper.vm.methodEmmitedOfPokemonOptions(196);
        
        expect(initialWrapper.vm.showImagePokemon).toBeTruthy();
        expect(initialWrapper.vm.messageResultGame).not.toBeNull();
    });
});
