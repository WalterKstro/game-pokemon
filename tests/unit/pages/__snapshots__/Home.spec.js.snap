// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for Home page Should create a correct snapshot of <Home/> 1`] = `
<div class="container-game">
  <h1 class="title">¿Quien es este Pokemon?</h1>
  <div class="container-main">
    <div class="container-picture"><img draggable="false" state="show" class="picture-pokemon picture-origin" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/196.svg" alt="pokemon"></div>
    <aside>
      <h2 class="list-title">Seleccione una opción</h2>
      <ul class="container-list">
        <li class="list-item"><button class="list-button" data-id="196">espeon</button></li>
        <li class="list-item"><button class="list-button" data-id="630">mandibuzz</button></li>
        <li class="list-item"><button class="list-button" data-id="173">cleffa</button></li>
        <li class="list-item"><button class="list-button" data-id="238">smoochum</button></li>
      </ul>
    </aside>
  </div>
  <!--v-if-->
</div>
`;
